src/cargo/core/compiler/mod.rs

```diff
@@ -1162,6 +1168,74 @@ fn features_args(unit: &Unit) -> Vec<OsString> {
     args
 }
 
+/// Generates the `--remap-path-scope` and `--remap-path-prefix` for [RFC 3127].
+/// See also unstable feature [`-Ztrim-paths`].
+///
+/// [RFC 3127]: https://rust-lang.github.io/rfcs/3127-trim-paths.html
+/// [`-Ztrim-paths`]: https://doc.rust-lang.org/nightly/cargo/reference/unstable.html#profile-trim-paths-option
+fn trim_paths_args(
+    cmd: &mut ProcessBuilder,
+    cx: &Context<'_, '_>,
+    unit: &Unit,
+    trim_paths: &TomlTrimPaths,
+) -> CargoResult<()> {
+    if trim_paths.is_none() {
+        return Ok(());
+    }
+
+    // feature gate was checked during mainfest/config parsing.
+    cmd.arg("-Zunstable-options");
+    cmd.arg(format!("-Zremap-path-scope={trim_paths}"));
+
+    let sysroot_remap = {
+        let sysroot = &cx.bcx.target_data.info(unit.kind).sysroot;
+        let mut remap = OsString::from("--remap-path-prefix=");
+        remap.push(sysroot);
+        remap.push("/lib/rustlib/src/rust"); // See also `detect_sysroot_src_path()`.
+        remap.push("=");
+        remap.push("/rustc/");
+        if let Some(commit_hash) = cx.bcx.rustc().commit_hash.as_ref() {
+            remap.push(commit_hash);
+        } else {
+            // This fallback aligns with rustc:
+            // <https://github.com/rust-lang/rust/blob/c2ef3516/src/bootstrap/src/lib.rs#L1113-L1116>
```

Comment:
Should this comment be on the whole if-statement?  I think this explains why we use the commit hash.